### PR TITLE
[Native][Issue-3593] Handle Task::updateBroadcastOutputBuffer failure gracefully with error log message

### PR DESCRIPTION
**Pull request by karteekmurthys**
The TaskManager may enter a race condition on Task::updateBrodacastBuffers() leading to an exception:"Output buffers not found" if the PartitionedOutputBuffers for that task are previously deleted by a downstream driver.

This PR handles this path gracefully by checking the error condition in a new `PartitionedOutputBuffersManager::getBufferIfExists()` API instead of `getBuffer()` call (which throws that exception).  An error message is also logged in this case.

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/TaskManager.cpp [hunk 1]
```diff
@@ -335,10 +335,20 @@ std::unique_ptr<TaskInfo> TaskManager::createOrUpdateTask(
 
   // TODO Handle possible race condition. Refer:
   // https://github.com/facebookincubator/velox/issues/3593
-  if (outputBuffers.type == protocol::BufferType::BROADCAST) {
-    execTask->updateBroadcastOutputBuffers(
-        outputBuffers.buffers.size(), outputBuffers.noMoreBufferIds);
-  }
+  if (auto ret =
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/TaskManager.cpp:338]: Why is the condition BufferType::BROADCAST removed ? That doesn't seem right to me.

File: presto-native-execution/presto_cpp/main/TaskManager.h [hunk 2]
```diff
@@ -103,6 +103,23 @@ class TaskManager {
       std::optional<protocol::Duration> maxWait,
       std::shared_ptr<http::CallbackRequestHandlerState> state);
 
+  /// Update buffer count for a given task. Returns nullptr, if update was
+  /// successful, else returns a std::unique_ptr to TaskInfo. Note: This method
+  /// is made public for unit testing purpose only.
+  inline std::unique_ptr<protocol::TaskInfo> updateBroadcastOutputBuffers(
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/TaskManager.h:109]: Is it possible to return and test a boolean for this method instead ? Returning nullptr for a successful update vs TaskInfo for a failure seems weird.

File: presto-native-execution/presto_cpp/main/TaskManager.cpp [hunk 3]
```diff
@@ -333,11 +333,8 @@ std::unique_ptr<TaskInfo> TaskManager::createOrUpdateTask(
 
   getDataForResultRequests(resultRequests);
 
-  // TODO Handle possible race condition. Refer:
-  // https://github.com/facebookincubator/velox/issues/3593
-  if (outputBuffers.type == protocol::BufferType::BROADCAST) {
-    execTask->updateBroadcastOutputBuffers(
-        outputBuffers.buffers.size(), outputBuffers.noMoreBufferIds);
+  if (!updateBroadcastOutputBuffers(prestoTask, execTask, outputBuffers)) {
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/TaskManager.cpp:336]: Why are you always calling updateBroadcastOutputBuffers ? It seems more readable to check the BROADCAST condition here than in the function called ?

File: presto-native-execution/presto_cpp/main/TaskManager.h [hunk 4]
```diff
@@ -103,6 +103,21 @@ class TaskManager {
       std::optional<protocol::Duration> maxWait,
       std::shared_ptr<http::CallbackRequestHandlerState> state);
 
+  /// Update buffer count for a given task for broadCast case.
+  /// Returns true, if update was successful, else returns false.
+  /// Also, returns true for a no-op case.
+  /// Note: This method is made public for unit testing purpose only.
+  inline bool updateBroadcastOutputBuffers(
+      std::shared_ptr<PrestoTask>& prestoTask,
```
Comment (aditi-pandit) [inline presto-native-execution/presto_cpp/main/TaskManager.h:111]: It doesn't seem like you are using prestoTask in this function. Why are you passing it ?

**Comment by v-jizhang** [comment]
@prestobot kick off tests

**Comment by karteekmurthys** [comment]
@mbasmanova @aditi-pandit would you please review this.

**Comment by karteekmurthys** [comment]
@mbasmanova @aditi-pandit would you please review this.

File: presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp [hunk 5]
```diff
@@ -917,5 +917,4 @@ TEST_F(TaskManagerTest, getResultsErrorPropagation) {
           .get(),
       VeloxException);
 }
-
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp:920]: Unrelated change?

**Review by mbasmanova — COMMENTED** [review]
@karteekmurthys The change looks good to me. Thanks. Please, update commit message to provide more details about this change.

> No testing required as we are not taking any action for error case.

Not sure if this is a good reason for not adding a test.

**Review by mbasmanova — APPROVED** [review]

**Review by aditi-pandit — APPROVED** [review]

**Comment by aditi-pandit** [comment]
> @karteekmurthys The change looks good to me. Thanks. Please, update commit message to provide more details about this change.
> 
> > No testing required as we are not taking any action for error case.
> 
> Not sure if this is a good reason for not adding a test.

+1. Your change has an effect on the behavior. It would be good to mock and check that you caught the error. Please try to address that in another PR.
